add Eio_unix.Sockopt for setting/getting socket options

avsm opened the pull request:
There are some socket options that are Linux-specific which are quite handy to have.  This adds a Eio_unix.Sockopt module which works on Eio_unix.Fd.t values.

This is a draft PR to check on the interface: the current PR replaces the need for Unix.setsockopt and has a single simple GADT for the socket options.  Alternatively we could just expose the extended ones in the same style as upstream OCaml as a separate set of types, too.  Opinions welcome.

anmonteiro commented on lib_eio/unix/eio_unix.mli:
```diff
@@ -84,6 +84,20 @@ module Stdenv : sig
+   | TCP_CORK : int t
+   | TCP_KEEPCNT : int t
+   | TCP_KEEPIDLE : int t
+   | TCP_KEEPINTVL : int t
```
could you also include TCP_NODELAY?

avsm replied:
Yeah, I'll add the complete set and tests once we decide on the interface.

anmonteiro replied:
On the interface: seems easy to use and I like the GADT approach for the options themselves.

I really like having just a single function to set socket options.

avsm replied:
added TCP_NODELAY

talex5 commented on lib_eio/unix/eio_unix.mli:
```diff
@@ -84,6 +84,20 @@ module Stdenv : sig
       You can use the functions in {!Eio.Stdenv} to access these fields if you prefer. *)
 end
 
+module Sockopt : sig
```
Would probably be better as `Eio_unix.Net.Sockopt`.

avsm replied:
Done in 92a06fe38597fae48c2f520d2d7d070e62cf23a3

talex5 commented on lib_eio/unix/eio_unix.mli:
```diff
@@ -84,6 +84,20 @@ module Stdenv : sig
 end
 
+module Sockopt : sig
+  type 'a t = 'a Eio_unix__Sockopt.t =
```
Why refer to the `Eio_unix__Sockopt` internal module here?

avsm replied:
Removed in latest patchset; all one extensible variant now.

talex5 reviewed (commented):
Seems like socket options could be useful on non-Unix systems too. We could define the type in Eio.Net as an extensible variant and put the common ones there, adding Unix specific ones in Eio_unix.Net, etc. Would have to decide what to do about unsupported options.

avsm commented:
> Seems like socket options could be useful on non-Unix systems too. We could define the type in Eio.Net as an extensible variant and put the common ones there, adding Unix specific ones in Eio_unix.Net, etc. Would have to decide what to do about unsupported options.

SGTM.  Right now, if a given sockopt isn't defined, it becomes (-1) and a call to it raises EINVAL (similar behaviour to  the OCaml Unix module).  

I'll move the common definitions that work on Windows as well to Eio.Net, and then put the remaining ones in Eio_unix.

talex5 commented:
In #322, @create2000 is asking about helping to finish this. Are we happy with the current design? What needs to happen beyond rebasing it?

It looks to me like some of the options (those not specific to Unix) should move to `Eio.Net` (e.g. `TCP_NODELAY` would also make sense in a unikernel). Possibly all the options should move there; it's OK to talk about an option, even if it's not supported.

Possibly `eio_unix` should define e.g. `type _ t += Unix_bool of Unix.socket_bool_option : bool t`, so all existing OCaml options can be used.

Might need a way to pass a list of options when connecting or binding (since we do that in one step, and maybe it's too late to set some options by then?). Might want some fancy heterogeneous list to hold them, since they have various types.

@create2000 I don't know if it's a good first issue or not. It uses some fancy types and involves C code. On the other hand, it's mostly done!

patricoferris commented:
Re:design -- my impression is that we should definitely move parts to `Eio.Net` that are portable, and in this instance perhaps all of the options and add the `Unix` version to allow for all existing OCaml options to _just work_. I think we will want to be able to get/set after connecting/binding anyway so this seems like a reasonable chunk of work to have in a single PR (basically what is already here, shuffled around a little)?

Perhaps a follow up PR can work to support options being set at the point of connecting/binding (as you mentioned, this seems very related to #713)?

avsm committed:
eio_linux: Add Linux-specific socket options

Co-authored-by: Thomas Leonard <talex5@gmail.com>

avsm commented:
I've pushed a rebase with all the suggestions here. There are loads of Linux and Windows specific ones to add, but I wanted to check if this is version is heading in the right direction first @talex5 @patricoferris

talex5 reviewed (commented):
This approach looks good to me.

avsm commented:
I've added a bunch of Linux specific calls now (skipping multicast for now, and some of the more exotic fast connect and tcp_info ones), and also dropped SO_ACCEPTCON as it's not portable to macOS with no reasonable alternative (have to do a listen and wait for an EINVAL). It can still be passed directly via the Unix passthrough, but it's not much use in Eio where you know the state of your sockets anyway.

talex5 commented on lib_eio_linux/eio_stubs.c:
```diff
@@ -258,3 +259,26 @@ CAMLprim value caml_eio_sockopt_int_get(value v_fd, value v_level, value v_optio
+  CAMLparam4(v_fd, v_level, v_option, v_val);
+  const char *str = String_val(v_val);
+  int ret = setsockopt(Int_val(v_fd), Int_val(v_level), Int_val(v_option),
+                       str, strlen(str) + 1);
```
Should probably  use `caml_string_length` to avoid problems with zero bytes.

talex5 commented on lib_eio_linux/eio_stubs.c:
```diff
@@ -258,3 +259,26 @@ CAMLprim value caml_eio_sockopt_int_get(value v_fd, value v_level, value v_optio
+  int ret = getsockopt(Int_val(v_fd), Int_val(v_level), Int_val(v_option),
+                       buffer, &len);
+  if (ret == -1) uerror("getsockopt", Nothing);
+  buffer[len < sizeof(buffer) ? len : sizeof(buffer) - 1] = '\0';
```
Likewise, I guess this should use `caml_alloc_initialized_string`.

avsm commented:
Squashed commits, should be ready for merge.

talex5 reviewed (commented):
(only reviewed the mock bit so far)

avsm commented on lib_eio/mock/sockopt.ml:
```diff
@@ -0,0 +1,47 @@
+
+let setsockopt : type a. string -> a Eio.Net.Sockopt.t -> a -> unit = fun label opt v ->
+  let opt_desc = pp_sockopt_value opt v in
+  traceln "%s: setsockopt %s" label opt_desc
```
fixed in a44bdb54bd832008721aba1681cc4ea57d69ede2

avsm commented on lib_eio/mock/sockopt.ml:
```diff
@@ -0,0 +1,47 @@
+  | Eio.Net.Sockopt.SO_LINGER -> traceln "%s: getsockopt SO_LINGER = None" label; None
+  | Eio.Net.Sockopt.SO_RCVTIMEO -> traceln "%s: getsockopt SO_RCVTIMEO = 0.0" label; 0.0
+  | Eio.Net.Sockopt.SO_SNDTIMEO -> traceln "%s: getsockopt SO_SNDTIMEO = 0.0" label; 0.0
+  | _ -> raise (Failure "Mock getsockopt not implemented")
```
I've left gets logging as I'd like to (in the future) have the mock layer remember the setsockopt values in a heterogenous map. I have lifted out the pp functions to be exposed in the main interface in a44bdb54bd832008721aba1681cc4ea57d69ede2 as that seems obviously useful

avsm commented:
Rebased to trunk, and some minor fixes also added while re-reviewing it.

avsm commented:
@talex5 @patricoferris I'm happy with the low-level implementation here, but I'm not sure if the Pi integration is right. We've got a new SOCKET module type now, and I think it won't be too onerous to mock, but I'm not sure...

avsm commented:
Need to ensure this is compatible with whatever approach we pick for #713 as well

talex5 commented:
I've merged some of this separately as #858 and #859 and rebased the remaining bit (Linux options) on them.

I then added another commit with some suggested changes:

- Move all options to `Eio.Net.Sockopt` so you can try them, even if the backend might not be Linux.
- Simplify code and tests a bit.

avsm commented on lib_eio/net.mli:
```diff
@@ -156,8 +161,74 @@ module Sockopt : sig
+    | IP_BIND_ADDRESS_NO_PORT : bool t
+    (** Inform the kernel to not reserve an ephemeral port when using bind()
+        with a port number of 0. The port will be chosen at connect() time. *)
+    | IP_LOCAL_PORT_RANGE : (int * int) t
```
This one needs a design opinion: I didn't wrap it in option as I felt that passing 0,0 is fine to reset it, but it can be done (at the cost of more complexity)

talex5 replied:
I agree; I don't think it's worth it.

talex5 commented on lib_eio_linux/tests/network.md:
```diff
@@ -0,0 +1,196 @@
+  try_setsockopt client Sockopt.TCP_KEEPCNT 5;
+  try_setsockopt client Sockopt.TCP_LINGER2 (Some 110);
+
+  (*
+  try_getsockopt client Sockopt.TCP_CONGESTION;
+  (* Try to set cubic if available - may fail based on system config *)
+  (try
+    try_setsockopt client Sockopt.TCP_CONGESTION "cubic";
+  with _ ->
+    traceln "Could not change TCP_CONGESTION (normal for unprivileged)");
+  *)
```
I disabled this test, as the result depends on the host configuration (as noted in the comment).

talex5 commented on lib_eio_linux/eio_stubs.c:
```diff
@@ -246,3 +249,45 @@ CAMLprim value caml_eio_clone3(value v_errors, value v_actions) {
+  int ret = getsockopt(Int_val(v_fd), Int_val(v_level), Int_val(v_option),
+                       buffer, &len);
+  if (ret == -1) uerror("getsockopt", Nothing);
+  v_result = caml_copy_string(buffer);
```
I added `strnlen` here - in theory the result might not have a null terminator if there wasn't room in the buffer. Though since `TCP_CA_NAME_MAX` is 16 we'll never hit this problem with the current set of options.

avsm commented on lib_eio_linux/eio_stubs.c:
```diff
@@ -246,3 +249,45 @@ CAMLprim value caml_eio_clone3(value v_errors, value v_actions) {
+  int ret = getsockopt(Int_val(v_fd), Int_val(v_level), Int_val(v_option),
+                       buffer, &len);
+  if (ret == -1) uerror("getsockopt", Nothing);
+  v_result = caml_alloc_initialized_string(strnlen(buffer, len), buffer);
```
Nothing good comes from returning a truncated string; the caller wont be able to use it meaningully. Instead of strnlen, why not check `len` after the getsockopt and raise an exception if it's the same as sizeof(buffer)?

talex5 replied:
Yes, could do. Not sure what exception to raise, though.

avsm replied:
Invalid_argument is reasonable here. It's a rare violation that indicates something's quite wrong, and not something expected to be caught and handled.

talex5 replied:
OK, done!
